File: pet_flutter/lib/services/signalr_service.dart
import 'dart:async';
import 'package:signalr_core/signalr_core.dart';
import '../config/api_config.dart';
import 'secure_storage.dart';

class SignalRService {
  static final SignalRService _instance = SignalRService._internal();
  factory SignalRService() => _instance;
  SignalRService._internal();

  HubConnection? _connection;
  bool _isConnected = false;
  String? _currentRoomKey;
  String? _currentUserId;
  String? _currentUserName;
  
  // Stream controllers for real-time events
  final StreamController<TimeSlotSelectionEvent> _timeSlotSelectedController = 
      StreamController<TimeSlotSelectionEvent>.broadcast();
  final StreamController<TimeSlotSelectionEvent> _timeSlotClearedController = 
      StreamController<TimeSlotSelectionEvent>.broadcast();
  final StreamController<bool> _connectionStatusController = 
      StreamController<bool>.broadcast();

  // Getters for streams
  Stream<TimeSlotSelectionEvent> get timeSlotSelectedStream => _timeSlotSelectedController.stream;
  Stream<TimeSlotSelectionEvent> get timeSlotClearedStream => _timeSlotClearedController.stream;
  Stream<bool> get connectionStatusStream => _connectionStatusController.stream;

  bool get isConnected => _isConnected;
  String? get currentRoomKey => _currentRoomKey;

  // Initialize SignalR connection
  Future<bool> initialize() async {
    if (_connection != null && _isConnected) {

      return true;
    }

    try {

      // Get user info for authentication
      final storage = SecureStorageService();
      final token = await storage.readToken();
      final userId = await storage.readUserId();
      final userName = await storage.readUserName();
      
      _currentUserId = userId;
      _currentUserName = userName ?? 'Anonymous User';

      // Create connection
      _connection = HubConnectionBuilder()
          .withUrl(
            '${ApiConfig.baseUrl}/timeSlotHub',
            HttpConnectionOptions(
              accessTokenFactory: () async => token,
              transport: HttpTransportType.webSockets,
              skipNegotiation: false,
            ),
          )
          .withAutomaticReconnect([0, 2000, 10000, 30000])
          .build();

      // Setup event handlers
      _setupEventHandlers();

      // Start connection
      await _connection!.start();
      _isConnected = true;
      _connectionStatusController.add(true);

      return true;
    } catch (e) {

      _isConnected = false;
      _connectionStatusController.add(false);
      return false;
    }
  }

  // Setup event handlers
  void _setupEventHandlers() {
    if (_connection == null) return;

    // Connection state events
    _connection!.onreconnecting((error) {

      _isConnected = false;
      _connectionStatusController.add(false);
    });

    _connection!.onreconnected((connectionId) {

      _isConnected = true;
      _connectionStatusController.add(true);
      
      // Rejoin current room if any
      if (_currentRoomKey != null) {
        joinTimeSlotRoom(_currentRoomKey!);
      }
    });

    _connection!.onclose((error) {

      _isConnected = false;
      _connectionStatusController.add(false);
    });

    // Time slot selection events
    _connection!.on('TimeSlotSelected', (arguments) {
      if (arguments != null && arguments.isNotEmpty) {
        try {
          final data = arguments[0] as Map<String, dynamic>;

          final event = TimeSlotSelectionEvent.fromJson(data);
          
          // Only process events from other users
          if (event.userId != _currentUserId) {

            _timeSlotSelectedController.add(event);
          } else {

          }
        } catch (e) {

        }
      }
    });

    _connection!.on('TimeSlotCleared', (arguments) {
      if (arguments != null && arguments.isNotEmpty) {
        try {
          final data = arguments[0] as Map<String, dynamic>;

          final event = TimeSlotSelectionEvent.fromJson(data);
          
          // Only process events from other users
          if (event.userId != _currentUserId) {

            _timeSlotClearedController.add(event);
          } else {

          }
        } catch (e) {

        }
      }
    });

    _connection!.on('UserJoinedTimeSlotRoom', (arguments) {
      if (arguments != null && arguments.isNotEmpty) {
        final data = arguments[0] as Map<String, dynamic>;

      }
    });

    _connection!.on('UserLeftTimeSlotRoom', (arguments) {
      if (arguments != null && arguments.isNotEmpty) {
        final data = arguments[0] as Map<String, dynamic>;

      }
    });
  }

  // Join time slot room for real-time updates
  Future<bool> joinTimeSlotRoom(String roomKey) async {
    if (!_isConnected || _connection == null) {

      return false;
    }

    try {
      // Leave current room first
      if (_currentRoomKey != null && _currentRoomKey != roomKey) {
        await leaveTimeSlotRoom(_currentRoomKey!);
      }

      await _connection!.invoke('JoinTimeSlotRoom', args: [roomKey]);
      _currentRoomKey = roomKey;

      return true;
    } catch (e) {

      return false;
    }
  }

  // Leave time slot room
  Future<bool> leaveTimeSlotRoom(String roomKey) async {
    if (!_isConnected || _connection == null) {
      return false;
    }

    try {
      await _connection!.invoke('LeaveTimeSlotRoom', args: [roomKey]);
      if (_currentRoomKey == roomKey) {
        _currentRoomKey = null;
      }

      return true;
    } catch (e) {

      return false;
    }
  }

  // Notify when user selects a time slot
  Future<bool> notifyTimeSlotSelected({
    required String roomKey,
    required String timeSlot,
    required String serviceId,
    required String staffId,
    required String date,
  }) async {
    if (!_isConnected || _connection == null) {

      return false;
    }

    try {
      // Parse serviceId and staffId to int for backend
      final serviceIdInt = int.tryParse(serviceId);
      final staffIdInt = int.tryParse(staffId);

      // Format data exactly as backend expects - PascalCase to match C# properties
      final data = {
        'RoomKey': roomKey,  // Capital R to match TimeSlotSelectionRequest
        'TimeSlot': timeSlot,  // Capital T to match TimeSlotSelectionRequest
        'UserId': _currentUserId ?? 'anonymous',  // Capital U
        'UserName': _currentUserName ?? 'Anonymous User',  // Capital U
        'ServiceId': serviceIdInt,  // int? type - parsed from string
        'StaffId': staffIdInt,  // int? type - parsed from string
        'Date': date,  // Capital D
      };

      // Send as single argument (the data object)
      await _connection!.invoke('NotifyTimeSlotSelected', args: [data]);

      return true;
    } catch (e) {

      return false;
    }
  }

  // Notify when user clears/deselects a time slot
  Future<bool> notifyTimeSlotCleared({
    required String roomKey,
    required String timeSlot,
    required String serviceId,
    required String staffId,
    required String date,
  }) async {
    if (!_isConnected || _connection == null) {

      return false;
    }

    try {
      // Parse serviceId and staffId to int for backend
      final serviceIdInt = int.tryParse(serviceId);
      final staffIdInt = int.tryParse(staffId);

      // Format data exactly as backend expects - PascalCase to match C# properties
      final data = {
        'RoomKey': roomKey,  // Capital R to match TimeSlotSelectionRequest
        'TimeSlot': timeSlot,  // Capital T to match TimeSlotSelectionRequest
        'UserId': _currentUserId ?? 'anonymous',  // Capital U
        'UserName': _currentUserName ?? 'Anonymous User',  // Capital U
        'ServiceId': serviceIdInt,  // int? type - parsed from string
        'StaffId': staffIdInt,  // int? type - parsed from string
        'Date': date,  // Capital D
      };

      // Send as single argument (the data object)
      await _connection!.invoke('NotifyTimeSlotCleared', args: [data]);

      return true;
    } catch (e) {

      return false;
    }
  }

  // Generate room key (same format as web)
  String generateRoomKey({
    required String serviceId,
    required String staffId,
    required String date,
  }) {
    return 'service_${serviceId}_staff_${staffId}_date_$date';
  }

  // Disconnect
  Future<void> disconnect() async {
    try {
      if (_currentRoomKey != null) {
        await leaveTimeSlotRoom(_currentRoomKey!);
      }
      
      await _connection?.stop();
      _connection = null;
      _isConnected = false;
      _currentRoomKey = null;
      _connectionStatusController.add(false);

    } catch (e) {

    }
  }

  // Dispose resources
  void dispose() {
    _timeSlotSelectedController.close();
    _timeSlotClearedController.close();
    _connectionStatusController.close();
    disconnect();
  }
}

// Event model for time slot selection
class TimeSlotSelectionEvent {
  final String roomKey;
  final String timeSlot;
  final String userId;
  final String userName;
  final String serviceId;
  final String staffId;
  final String date;
  final DateTime timestamp;

  TimeSlotSelectionEvent({
    required this.roomKey,
    required this.timeSlot,
    required this.userId,
    required this.userName,
    required this.serviceId,
    required this.staffId,
    required this.date,
    required this.timestamp,
  });

  factory TimeSlotSelectionEvent.fromJson(Map<String, dynamic> json) {
    return TimeSlotSelectionEvent(
      // Backend sends PascalCase (TimeSlot, UserId, etc.)
      roomKey: json['RoomKey']?.toString() ?? json['roomKey']?.toString() ?? '',
      timeSlot: json['TimeSlot']?.toString() ?? json['timeSlot']?.toString() ?? '',
      userId: json['UserId']?.toString() ?? json['userId']?.toString() ?? '',
      userName: json['UserName']?.toString() ?? json['userName']?.toString() ?? 'Unknown User',
      // Backend sends int for serviceId and staffId, convert to string
      serviceId: json['ServiceId']?.toString() ?? json['serviceId']?.toString() ?? '',
      staffId: json['StaffId']?.toString() ?? json['staffId']?.toString() ?? '',
      date: json['Date']?.toString() ?? json['date']?.toString() ?? '',
      timestamp: DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'roomKey': roomKey,
      'timeSlot': timeSlot,
      'userId': userId,
      'userName': userName,
      'serviceId': serviceId,
      'staffId': staffId,
      'date': date,
      'timestamp': timestamp.toIso8601String(),
    };
  }

  @override
  String toString() {
    return 'TimeSlotSelectionEvent(timeSlot: $timeSlot, userName: $userName)';
  }
}

File: pet_flutter/pet_web/src/services/signalrService.js
import * as signalR from '@microsoft/signalr';
import { notification } from 'antd';

class SignalRService {
  constructor() {
    this.connection = null;
    this.currentServiceGroup = null;
    this.reservations = new Map(); // Track temporary reservations
    this.isConnected = false;
    this.listeners = new Map(); // Event listeners
  }

  // Initialize connection
  async initialize() {
    if (this.connection && this.connection.state === 'Connected') {
      console.log('SignalR already connected');
      return this.connection;
    }

    // Always create fresh connection to avoid state issues
    await this.disconnect();
    this.connection = null;

    try {
      console.log('🔌 Creating fresh SignalR connection...');
      
      // Create connection to TimeSlotSharingHub for real-time slot sharing
      this.connection = new signalR.HubConnectionBuilder()
        .withUrl(`${process.env.REACT_APP_BASE_URL || "https://bepetwebapi20260223122715-hsfwcberazegd0hd.southeastasia-01.azurewebsites.net"}/timeSlotHub`, {
          skipNegotiation: false,
          transport: signalR.HttpTransportType.LongPolling | signalR.HttpTransportType.WebSockets
        })
        .withAutomaticReconnect([0, 2000, 10000, 30000])
        .configureLogging(signalR.LogLevel.Information)
        .build();

      // Set up event handlers
      this.setupEventHandlers();

      console.log('🔌 Starting SignalR connection...');
      await this.connection.start();
      this.isConnected = true;
      console.log('✅ SignalR Connected successfully, state:', this.connection.state);
      
      // Notify listeners about connection
      this.notifyListeners('connected', true);
      
      return this.connection;
    } catch (err) {
      console.error('❌ SignalR Connection Error:', err);
      this.isConnected = false;
      
      // Reset connection on error
      this.connection = null;
      
      throw err;
    }
  }



  // Setup event handlers
  setupEventHandlers() {
    if (!this.connection) return;

    // Connection events
    this.connection.onreconnecting(() => {
      this.isConnected = false;
      console.log('SignalR Reconnecting...');
      this.notifyListeners('reconnecting', true);
    });

    this.connection.onreconnected(() => {
      this.isConnected = true;
      console.log('SignalR Reconnected');
      this.notifyListeners('reconnected', true);
      
      // Rejoin current group if any
      if (this.currentServiceGroup) {
        this.joinServiceGroup(
          this.currentServiceGroup.serviceId, 
          this.currentServiceGroup.date, 
          this.currentServiceGroup.staffId
        );
      }
    });

    this.connection.onclose(() => {
      this.isConnected = false;
      console.log('SignalR Disconnected');
      this.notifyListeners('disconnected', true);
    });

    // ===== TIME SLOT SELECTION SHARING EVENTS (NEW) =====
    this.connection.on('TimeSlotSelected', (data) => {
      console.log('🔔 [SignalR] TimeSlotSelected received:', data);
      this.notifyListeners('timeSlotSelected', data);
    });

    this.connection.on('TimeSlotCleared', (data) => {
      console.log('🔔 [SignalR] TimeSlotCleared received:', data);
      this.notifyListeners('timeSlotCleared', data);
    });

    this.connection.on('UserJoinedTimeSlotRoom', (data) => {
      console.log('🔔 [SignalR] UserJoinedTimeSlotRoom:', data);
      this.notifyListeners('userJoinedTimeSlotRoom', data);
    });

    this.connection.on('UserLeftTimeSlotRoom', (data) => {
      console.log('🔔 [SignalR] UserLeftTimeSlotRoom:', data);
      this.notifyListeners('userLeftTimeSlotRoom', data);
    });

    // ===== SLOT RESERVATION EVENTS (EXISTING) =====
    this.connection.on('ReservationSuccess', (data) => {
      console.log('Reservation Success:', data);
      
      // Store our own reservation
      this.reservations.set(this.createSlotKey(data.slotStartTime), {
        ...data,
        isOwnReservation: true
      });

      this.notifyListeners('reservationSuccess', data);
      
      notification.success({
        message: 'Đặt chỗ thành công',
        description: `Bạn đã giữ chỗ khung giờ ${new Date(data.slotStartTime).toLocaleTimeString('vi-VN')}`,
        duration: 3
      });
    });

    this.connection.on('ReservationFailed', (data) => {
      console.log('Reservation Failed:', data);
      this.notifyListeners('reservationFailed', data);
      
      notification.error({
        message: 'Không thể đặt chỗ',
        description: data.message,
        duration: 4
      });
    });

    this.connection.on('SlotReserved', (data) => {
      console.log('Slot Reserved by others:', data);
      
      // Store other's reservation
      this.reservations.set(this.createSlotKey(data.slotStartTime), {
        ...data,
        isOwnReservation: false
      });

      this.notifyListeners('slotReserved', data);
      
      notification.info({
        message: 'Khung giờ đã được đặt',
        description: `${data.reservedBy} đã chọn khung giờ ${new Date(data.slotStartTime).toLocaleTimeString('vi-VN')}`,
        duration: 3
      });
    });

    this.connection.on('SlotReleased', (data) => {
      console.log('Slot Released:', data);
      
      // Remove reservation
      this.reservations.delete(this.createSlotKey(data.slotStartTime));

      this.notifyListeners('slotReleased', data);
      
      if (data.reason === 'expired') {
        notification.warning({
          message: 'Khung giờ đã được giải phóng',
          description: `Khung giờ ${new Date(data.slotStartTime).toLocaleTimeString('vi-VN')} đã hết thời gian giữ chỗ`,
          duration: 3
        });
      }
    });

    this.connection.on('CurrentReservations', (reservations) => {
      console.log('Current Reservations:', reservations);
      
      // Update reservations map
      this.reservations.clear();
      reservations.forEach(reservation => {
        this.reservations.set(
          this.createSlotKey(reservation.slotStartTime), 
          reservation
        );
      });

      this.notifyListeners('currentReservations', reservations);
    });

    this.connection.on('ReservationExtended', (data) => {
      console.log('Reservation Extended:', data);
      
      // Update reservation expiry
      const key = this.createSlotKey(data.slotStartTime);
      const reservation = this.reservations.get(key);
      if (reservation) {
        reservation.expiresAt = data.newExpiresAt;
      }

      this.notifyListeners('reservationExtended', data);
      
      notification.success({
        message: 'Gia hạn thành công',
        description: `Đã gia hạn khung giờ ${new Date(data.slotStartTime).toLocaleTimeString('vi-VN')}`,
        duration: 3
      });
    });
  }

  // Helper to create unique slot key
  createSlotKey(slotStartTime) {
    return new Date(slotStartTime).getTime().toString();
  }

  // Join service group to receive updates
  async joinServiceGroup(serviceId, date, staffId = 'any') {
    if (!this.isConnected || !this.connection) {
      console.error('SignalR not connected');
      return false;
    }

    try {
      await this.connection.invoke('JoinServiceGroup', 
        serviceId.toString(), 
        date, 
        staffId?.toString() || 'any'
      );
      
      this.currentServiceGroup = { serviceId, date, staffId };
      console.log(`Joined service group: ${serviceId}_${date}_${staffId || 'any'}`);
      return true;
    } catch (err) {
      console.error('Error joining service group:', err);
      return false;
    }
  }

  // Leave service group
  async leaveServiceGroup() {
    if (!this.isConnected || !this.connection || !this.currentServiceGroup) {
      return false;
    }

    try {
      const { serviceId, date, staffId } = this.currentServiceGroup;
      await this.connection.invoke('LeaveServiceGroup', 
        serviceId.toString(), 
        date, 
        staffId?.toString() || 'any'
      );
      
      this.currentServiceGroup = null;
      this.reservations.clear();
      console.log('Left service group');
      return true;
    } catch (err) {
      console.error('Error leaving service group:', err);
      return false;
    }
  }

  // Reserve a slot
  async reserveSlot(slotData) {
    if (!this.isConnected || !this.connection) {
      console.error('SignalR not connected');
      return false;
    }

    try {
      await this.connection.invoke('ReserveSlot', {
        serviceId: slotData.serviceId,
        staffId: slotData.staffId,
        startTime: slotData.startTime,
        endTime: slotData.endTime,
        date: slotData.date,
        userName: slotData.userName || 'Khách hàng'
      });
      return true;
    } catch (err) {
      console.error('Error reserving slot:', err);
      return false;
    }
  }

  // Release a slot
  async releaseSlot(slotData) {
    if (!this.isConnected || !this.connection) {
      console.error('SignalR not connected');
      return false;
    }

    try {
      await this.connection.invoke('ReleaseSlot', {
        serviceId: slotData.serviceId,
        staffId: slotData.staffId,
        startTime: slotData.startTime,
        date: slotData.date
      });
      return true;
    } catch (err) {
      console.error('Error releasing slot:', err);
      return false;
    }
  }

  // Extend reservation
  async extendReservation(slotData) {
    if (!this.isConnected || !this.connection) {
      console.error('SignalR not connected');
      return false;
    }

    try {
      await this.connection.invoke('ExtendReservation', {
        serviceId: slotData.serviceId,
        startTime: slotData.startTime
      });
      return true;
    } catch (err) {
      console.error('Error extending reservation:', err);
      return false;
    }
  }

  // Check if slot is reserved
  isSlotReserved(slotStartTime) {
    const key = this.createSlotKey(slotStartTime);
    return this.reservations.has(key);
  }

  // Get reservation info for slot
  getSlotReservation(slotStartTime) {
    const key = this.createSlotKey(slotStartTime);
    return this.reservations.get(key);
  }

  // Get all current reservations
  getAllReservations() {
    return Array.from(this.reservations.values());
  }

  // Add event listener
  addEventListener(event, callback) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event).add(callback);
  }

  // Remove event listener
  removeEventListener(event, callback) {
    if (this.listeners.has(event)) {
      this.listeners.get(event).delete(callback);
    }
  }

  // Notify listeners
  notifyListeners(event, data) {
    if (this.listeners.has(event)) {
      this.listeners.get(event).forEach(callback => {
        try {
          callback(data);
        } catch (err) {
          console.error('Error in event listener:', err);
        }
      });
    }
  }

  // Disconnect
  async disconnect() {
    if (this.connection) {
      try {
        await this.leaveServiceGroup();
        await this.connection.stop();
      } catch (err) {
        console.error('Error disconnecting:', err);
      } finally {
        this.connection = null;
        this.isConnected = false;
        this.currentServiceGroup = null;
        this.reservations.clear();
        this.listeners.clear();
      }
    }
  }

  // Get connection status
  getConnectionStatus() {
    return {
      isConnected: this.isConnected,
      currentGroup: this.currentServiceGroup,
      reservationsCount: this.reservations.size
    };
  }

  // ===== TIME SLOT SELECTION SHARING METHODS (NEW) =====
  
  // Join time slot room for real-time selection sharing
  async joinTimeSlotRoom(roomKey) {
    if (!this.isConnected || !this.connection) {
      console.error('SignalR not connected');
      return false;
    }

    try {
      await this.connection.invoke('JoinTimeSlotRoom', roomKey);
      console.log(`✅ [SignalR] Joined time slot room: ${roomKey}`);
      return true;
    } catch (err) {
      console.error('❌ [SignalR] Error joining time slot room:', err);
      return false;
    }
  }

  // Leave time slot room
  async leaveTimeSlotRoom(roomKey) {
    if (!this.isConnected || !this.connection) {
      return false;
    }

    try {
      await this.connection.invoke('LeaveTimeSlotRoom', roomKey);
      console.log(`✅ [SignalR] Left time slot room: ${roomKey}`);
      return true;
    } catch (err) {
      console.error('❌ [SignalR] Error leaving time slot room:', err);
      return false;
    }
  }

  // Notify when user selects a time slot
  async notifyTimeSlotSelected(data) {
    if (!this.isConnected || !this.connection) {
      console.error('SignalR not connected');
      return false;
    }

    try {
      await this.connection.invoke('NotifyTimeSlotSelected', data);
      console.log(`📤 [SignalR] Notified time slot selected:`, data);
      return true;
    } catch (err) {
      console.error('❌ [SignalR] Error notifying time slot selected:', err);
      return false;
    }
  }

  // Notify when user clears/deselects a time slot
  async notifyTimeSlotCleared(data) {
    if (!this.isConnected || !this.connection) {
      console.error('SignalR not connected');
      return false;
    }

    try {
      await this.connection.invoke('NotifyTimeSlotCleared', data);
      console.log(`📤 [SignalR] Notified time slot cleared:`, data);
      return true;
    } catch (err) {
      console.error('❌ [SignalR] Error notifying time slot cleared:', err);
      return false;
    }
  }

  // Get current connection for direct use in components
  getConnection() {
    return this.connection;
  }

  // Check if connected
  isConnectionReady() {
    return this.isConnected && this.connection;
  }
}

// Export singleton instance
export default new SignalRService(); 